File: hello-edge-tts-java/src/main/java/com/example/hellotts/SSMLUtils.java
package com.example.hellotts;

import java.util.*;
import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.function.BiFunction;

/**
 * SSML (Speech Synthesis Markup Language) utilities for Edge TTS.
 * 
 * This class provides builder patterns and validation for creating
 * SSML markup for use with Microsoft Edge TTS service.
 */
public class SSMLUtils {

    /**
     * Builder for creating SSML markup
     */
    public static class SSMLBuilder {
        private final String voice;
        private final String lang;
        private final List<String> elements;

        public SSMLBuilder(String voice) {
            this.voice = voice;
            this.lang = extractLanguage(voice);
            this.elements = new ArrayList<>();
        }

        public SSMLBuilder(String voice, String lang) {
            this.voice = voice;
            this.lang = lang;
            this.elements = new ArrayList<>();
        }

        private static String extractLanguage(String voice) {
            String[] parts = voice.split("-");
            if (parts.length >= 2) {
                return parts[0] + "-" + parts[1];
            }
            return "en-US";
        }

        /**
         * Add plain text
         */
        public SSMLBuilder addText(String text) {
            elements.add(text);
            return this;
        }

        /**
         * Add text with prosody controls
         */
        public SSMLBuilder addProsody(String text, String rate, String pitch, String volume) {
            List<String> attrs = new ArrayList<>();
            if (rate != null) attrs.add("rate=\"" + rate + "\"");
            if (pitch != null) attrs.add("pitch=\"" + pitch + "\"");
            if (volume != null) attrs.add("volume=\"" + volume + "\"");

            String attrStr = attrs.isEmpty() ? "" : " " + String.join(" ", attrs);
            elements.add("<prosody" + attrStr + ">" + text + "</prosody>");
            return this;
        }

        /**
         * Add emphasized text
         */
        public SSMLBuilder addEmphasis(String text, String level) {
            elements.add("<emphasis level=\"" + level + "\">" + text + "</emphasis>");
            return this;
        }

        /**
         * Add a break/pause
         */
        public SSMLBuilder addBreak(String time) {
            elements.add("<break time=\"" + time + "\"/>");
            return this;
        }

        /**
         * Add say-as element for special text interpretation
         */
        public SSMLBuilder addSayAs(String text, String interpretAs, String format) {
            String formatAttr = format != null ? " format=\"" + format + "\"" : "";
            elements.add("<say-as interpret-as=\"" + interpretAs + "\"" + formatAttr + ">" + text + "</say-as>");
            return this;
        }

        /**
         * Add phoneme pronunciation
         */
        public SSMLBuilder addPhoneme(String text, String alphabet, String ph) {
            elements.add("<phoneme alphabet=\"" + alphabet + "\" ph=\"" + ph + "\">" + text + "</phoneme>");
            return this;
        }

        /**
         * Add substitution
         */
        public SSMLBuilder addSub(String text, String alias) {
            elements.add("<sub alias=\"" + alias + "\">" + text + "</sub>");
            return this;
        }

        /**
         * Build the complete SSML markup
         */
        public String build() {
            String content = String.join("", elements);
            return String.format(
                "<speak version=\"1.0\" xmlns=\"http://www.w3.org/2001/10/synthesis\" xml:lang=\"%s\">\n" +
                "    <voice name=\"%s\">\n" +
                "        %s\n" +
                "    </voice>\n" +
                "</speak>",
                lang, voice, content
            );
        }
    }

    /**
     * Validator for SSML markup
     */
    public static class SSMLValidator {
        private static final Set<String> VALID_PROSODY_RATES = Set.of(
            "x-slow", "slow", "medium", "fast", "x-fast"
        );

        private static final Set<String> VALID_PROSODY_PITCHES = Set.of(
            "x-low", "low", "medium", "high", "x-high"
        );

        private static final Set<String> VALID_PROSODY_VOLUMES = Set.of(
            "silent", "x-soft", "soft", "medium", "loud", "x-loud"
        );

        private static final Set<String> VALID_EMPHASIS_LEVELS = Set.of(
            "strong", "moderate", "reduced"
        );

        private static final Set<String> VALID_BREAK_STRENGTHS = Set.of(
            "none", "x-weak", "weak", "medium", "strong", "x-strong"
        );

        /**
         * Validate SSML markup and return list of errors
         */
        public static List<String> validate(String ssml) {
            List<String> errors = new ArrayList<>();

            // Basic validation
            if (!ssml.trim().startsWith("<speak")) {
                errors.add("SSML must start with <speak> element");
            }

            if (!ssml.contains("version=\"1.0\"")) {
                errors.add("Missing version=\"1.0\" attribute in <speak> element");
            }

            if (!ssml.contains("xmlns=\"http://www.w3.org/2001/10/synthesis\"")) {
                errors.add("Missing xmlns attribute in <speak> element");
            }

            // Validate specific elements
            validateProsodyElements(ssml, errors);
            validateEmphasisElements(ssml, errors);
            validateBreakElements(ssml, errors);

            return errors;
        }

        private static void validateProsodyElements(String ssml, List<String> errors) {
            Pattern prosodyPattern = Pattern.compile("<prosody\\s+([^>]+)>");
            Matcher matcher = prosodyPattern.matcher(ssml);

            while (matcher.find()) {
                String attrs = matcher.group(1);

                Pattern ratePattern = Pattern.compile("rate=\"([^\"]+)\"");
                Matcher rateMatch = ratePattern.matcher(attrs);
                if (rateMatch.find()) {
                    String rate = rateMatch.group(1);
                    if (!VALID_PROSODY_RATES.contains(rate) && 
                        !rate.endsWith("%") && !rate.endsWith("Hz")) {
                        errors.add("Invalid prosody rate: " + rate);
                    }
                }

                Pattern pitchPattern = Pattern.compile("pitch=\"([^\"]+)\"");
                Matcher pitchMatch = pitchPattern.matcher(attrs);
                if (pitchMatch.find()) {
                    String pitch = pitchMatch.group(1);
                    if (!VALID_PROSODY_PITCHES.contains(pitch) && 
                        !pitch.endsWith("Hz") && !pitch.endsWith("st")) {
                        errors.add("Invalid prosody pitch: " + pitch);
                    }
                }

                Pattern volumePattern = Pattern.compile("volume=\"([^\"]+)\"");
                Matcher volumeMatch = volumePattern.matcher(attrs);
                if (volumeMatch.find()) {
                    String volume = volumeMatch.group(1);
                    if (!VALID_PROSODY_VOLUMES.contains(volume) && !volume.endsWith("dB")) {
                        errors.add("Invalid prosody volume: " + volume);
                    }
                }
            }
        }

        private static void validateEmphasisElements(String ssml, List<String> errors) {
            Pattern emphasisPattern = Pattern.compile("<emphasis\\s+level=\"([^\"]+)\"");
            Matcher matcher = emphasisPattern.matcher(ssml);

            while (matcher.find()) {
                String level = matcher.group(1);
                if (!VALID_EMPHASIS_LEVELS.contains(level)) {
                    errors.add("Invalid emphasis level: " + level);
                }
            }
        }

        private static void validateBreakElements(String ssml, List<String> errors) {
            Pattern breakPattern = Pattern.compile("<break\\s+([^>]+)/>");
            Matcher matcher = breakPattern.matcher(ssml);

            while (matcher.find()) {
                String attrs = matcher.group(1);

                Pattern timePattern = Pattern.compile("time=\"([^\"]+)\"");
                Matcher timeMatch = timePattern.matcher(attrs);
                if (timeMatch.find()) {
                    String time = timeMatch.group(1);
                    if (!time.endsWith("s") && !time.endsWith("ms")) {
                        errors.add("Invalid break time format: " + time);
                    }
                }

                Pattern strengthPattern = Pattern.compile("strength=\"([^\"]+)\"");
                Matcher strengthMatch = strengthPattern.matcher(attrs);
                if (strengthMatch.find()) {
                    String strength = strengthMatch.group(1);
                    if (!VALID_BREAK_STRENGTHS.contains(strength)) {
                        errors.add("Invalid break strength: " + strength);
                    }
                }
            }
        }
    }

    /**
     * Predefined SSML templates
     */
    public static class SSMLTemplates {
        private static final Map<String, BiFunction<String, String, String>> templates = Map.of(
            "slow_speech", (text, voice) -> new SSMLBuilder(voice).addProsody(text, "slow", null, null).build(),
            "fast_speech", (text, voice) -> new SSMLBuilder(voice).addProsody(text, "fast", null, null).build(),
            "whisper", (text, voice) -> new SSMLBuilder(voice).addProsody(text, "slow", null, "x-soft").build(),
            "excited", (text, voice) -> new SSMLBuilder(voice).addProsody(text, "fast", "high", "loud").build(),
            "calm", (text, voice) -> new SSMLBuilder(voice).addProsody(text, "slow", "low", "soft").build(),
            "emphasis_strong", (text, voice) -> new SSMLBuilder(voice).addEmphasis(text, "strong").build(),
            "with_pauses", (text, voice) -> {
                if (text.contains(".")) {
                    String[] parts = text.split("\\.", 2);
                    return new SSMLBuilder(voice)
                        .addText(parts[0])
                        .addBreak("1s")
                        .addText(parts.length > 1 ? parts[1] : "")
                        .build();
                }
                return new SSMLBuilder(voice).addText(text).build();
            }
        );

        /**
         * Create SSML using a predefined template
         */
        public static String createFromTemplate(String templateName, String text, String voice) {
            BiFunction<String, String, String> template = templates.get(templateName);
            if (template == null) {
                String available = String.join(", ", templates.keySet());
                throw new IllegalArgumentException("Unknown template '" + templateName + "'. Available: " + available);
            }
            return template.apply(text, voice);
        }

        /**
         * Get list of available template names
         */
        public static Set<String> getAvailableTemplates() {
            return templates.keySet();
        }
    }

    /**
     * Validate SSML markup
     */
    public static List<String> validateSSML(String ssml, boolean raiseOnError) {
        List<String> errors = SSMLValidator.validate(ssml);
        
        if (!errors.isEmpty() && raiseOnError) {
            throw new IllegalArgumentException("SSML validation failed: " + String.join("; ", errors));
        }
        
        return errors;
    }

    /**
     * Create SSML with prosody controls
     */
    public static String createSSML(String text, String voice, String rate, String pitch, String volume) {
        return new SSMLBuilder(voice).addProsody(text, rate, pitch, volume).build();
    }

    /**
     * Create SSML with emphasis
     */
    public static String createEmphasisSSML(String text, String voice, String level) {
        return new SSMLBuilder(voice).addEmphasis(text, level).build();
    }

    /**
     * Create SSML with breaks between text parts
     */
    public static String createBreakSSML(String[] textParts, String voice, String breakTime) {
        SSMLBuilder builder = new SSMLBuilder(voice);
        
        for (int i = 0; i < textParts.length; i++) {
            builder.addText(textParts[i]);
            if (i < textParts.length - 1) {
                builder.addBreak(breakTime);
            }
        }
        
        return builder.build();
    }
}

File: hello-edge-tts-rust/src/ssml_utils.rs


/// SSML (Speech Synthesis Markup Language) utilities for Edge TTS
/// 
/// This module provides builder patterns and validation for creating
/// SSML markup for use with Microsoft Edge TTS service.

/// Builder for creating SSML markup
pub struct SSMLBuilder {
    voice: String,
    lang: String,
    elements: Vec<String>,
}

impl SSMLBuilder {
    /// Create a new SSML builder
    pub fn new(voice: &str) -> Self {
        let lang = Self::extract_language(voice);
        Self {
            voice: voice.to_string(),
            lang,
            elements: Vec::new(),
        }
    }

    /// Create a new SSML builder with explicit language
    pub fn with_language(voice: &str, lang: &str) -> Self {
        Self {
            voice: voice.to_string(),
            lang: lang.to_string(),
            elements: Vec::new(),
        }
    }

    /// Extract language code from voice name
    fn extract_language(voice: &str) -> String {
        let parts: Vec<&str> = voice.split('-').collect();
        if parts.len() >= 2 {
            format!("{}-{}", parts[0], parts[1])
        } else {
            "en-US".to_string()
        }
    }

    /// Add plain text
    pub fn add_text(mut self, text: &str) -> Self {
        self.elements.push(text.to_string());
        self
    }

    /// Add text with prosody controls
    pub fn add_prosody(mut self, text: &str, rate: Option<&str>, pitch: Option<&str>, volume: Option<&str>) -> Self {
        let mut attrs = Vec::new();
        if let Some(r) = rate {
            attrs.push(format!("rate=\"{}\"", r));
        }
        if let Some(p) = pitch {
            attrs.push(format!("pitch=\"{}\"", p));
        }
        if let Some(v) = volume {
            attrs.push(format!("volume=\"{}\"", v));
        }

        let attr_str = if attrs.is_empty() {
            String::new()
        } else {
            format!(" {}", attrs.join(" "))
        };

        self.elements.push(format!("<prosody{}>{}</prosody>", attr_str, text));
        self
    }

    /// Add emphasized text
    pub fn add_emphasis(mut self, text: &str, level: &str) -> Self {
        self.elements.push(format!("<emphasis level=\"{}\">{}</emphasis>", level, text));
        self
    }

    /// Add a break/pause
    pub fn add_break(mut self, time: &str) -> Self {
        self.elements.push(format!("<break time=\"{}\"/>", time));
        self
    }

    /// Add say-as element for special text interpretation
    pub fn add_say_as(mut self, text: &str, interpret_as: &str, format: Option<&str>) -> Self {
        let format_attr = format.map(|f| format!(" format=\"{}\"", f)).unwrap_or_default();
        self.elements.push(format!("<say-as interpret-as=\"{}\"{}>{}</say-as>", interpret_as, format_attr, text));
        self
    }

    /// Add phoneme pronunciation
    pub fn add_phoneme(mut self, text: &str, alphabet: &str, ph: &str) -> Self {
        self.elements.push(format!("<phoneme alphabet=\"{}\" ph=\"{}\">{}</phoneme>", alphabet, ph, text));
        self
    }

    /// Add substitution
    pub fn add_sub(mut self, text: &str, alias: &str) -> Self {
        self.elements.push(format!("<sub alias=\"{}\">{}</sub>", alias, text));
        self
    }

    /// Build the complete SSML markup
    pub fn build(self) -> String {
        let content = self.elements.join("");
        format!(
            r#"<speak version="1.0" xmlns="http://www.w3.org/2001/10/synthesis" xml:lang="{}">
    <voice name="{}">
        {}
    </voice>
</speak>"#,
            self.lang, self.voice, content
        )
    }
}

/// Validator for SSML markup
pub struct SSMLValidator;

impl SSMLValidator {
    const VALID_PROSODY_RATES: &'static [&'static str] = &[
        "x-slow", "slow", "medium", "fast", "x-fast"
    ];

    const VALID_PROSODY_PITCHES: &'static [&'static str] = &[
        "x-low", "low", "medium", "high", "x-high"
    ];

    const VALID_PROSODY_VOLUMES: &'static [&'static str] = &[
        "silent", "x-soft", "soft", "medium", "loud", "x-loud"
    ];

    const VALID_EMPHASIS_LEVELS: &'static [&'static str] = &[
        "strong", "moderate", "reduced"
    ];

    const VALID_BREAK_STRENGTHS: &'static [&'static str] = &[
        "none", "x-weak", "weak", "medium", "strong", "x-strong"
    ];

    /// Validate SSML markup and return list of errors
    pub fn validate(ssml: &str) -> Vec<String> {
        let mut errors = Vec::new();

        // Basic validation
        if !ssml.trim_start().starts_with("<speak") {
            errors.push("SSML must start with <speak> element".to_string());
        }

        if !ssml.contains("version=\"1.0\"") {
            errors.push("Missing version=\"1.0\" attribute in <speak> element".to_string());
        }

        if !ssml.contains("xmlns=\"http://www.w3.org/2001/10/synthesis\"") {
            errors.push("Missing xmlns attribute in <speak> element".to_string());
        }

        // Validate specific elements
        Self::validate_prosody_elements(ssml, &mut errors);
        Self::validate_emphasis_elements(ssml, &mut errors);
        Self::validate_break_elements(ssml, &mut errors);

        errors
    }

    fn validate_prosody_elements(ssml: &str, errors: &mut Vec<String>) {
        use regex::Regex;
        
        let prosody_regex = Regex::new(r"<prosody\s+([^>]+)>").unwrap();
        
        for caps in prosody_regex.captures_iter(ssml) {
            let attrs = &caps[1];
            
            if let Some(rate_caps) = Regex::new(r#"rate="([^"]+)""#).unwrap().captures(attrs) {
                let rate = &rate_caps[1];
                if !Self::VALID_PROSODY_RATES.contains(&rate) && 
                   !rate.ends_with('%') && !rate.ends_with("Hz") {
                    errors.push(format!("Invalid prosody rate: {}", rate));
                }
            }

            if let Some(pitch_caps) = Regex::new(r#"pitch="([^"]+)""#).unwrap().captures(attrs) {
                let pitch = &pitch_caps[1];
                if !Self::VALID_PROSODY_PITCHES.contains(&pitch) && 
                   !pitch.ends_with("Hz") && !pitch.ends_with("st") {
                    errors.push(format!("Invalid prosody pitch: {}", pitch));
                }
            }

            if let Some(volume_caps) = Regex::new(r#"volume="([^"]+)""#).unwrap().captures(attrs) {
                let volume = &volume_caps[1];
                if !Self::VALID_PROSODY_VOLUMES.contains(&volume) && !volume.ends_with("dB") {
                    errors.push(format!("Invalid prosody volume: {}", volume));
                }
            }
        }
    }

    fn validate_emphasis_elements(ssml: &str, errors: &mut Vec<String>) {
        use regex::Regex;
        
        let emphasis_regex = Regex::new(r#"<emphasis\s+level="([^"]+)""#).unwrap();
        
        for caps in emphasis_regex.captures_iter(ssml) {
            let level = &caps[1];
            if !Self::VALID_EMPHASIS_LEVELS.contains(&level) {
                errors.push(format!("Invalid emphasis level: {}", level));
            }
        }
    }

    fn validate_break_elements(ssml: &str, errors: &mut Vec<String>) {
        use regex::Regex;
        
        let break_regex = Regex::new(r"<break\s+([^>]+)/>").unwrap();
        
        for caps in break_regex.captures_iter(ssml) {
            let attrs = &caps[1];
            
            if let Some(time_caps) = Regex::new(r#"time="([^"]+)""#).unwrap().captures(attrs) {
                let time = &time_caps[1];
                if !time.ends_with('s') && !time.ends_with("ms") {
                    errors.push(format!("Invalid break time format: {}", time));
                }
            }

            if let Some(strength_caps) = Regex::new(r#"strength="([^"]+)""#).unwrap().captures(attrs) {
                let strength = &strength_caps[1];
                if !Self::VALID_BREAK_STRENGTHS.contains(&strength) {
                    errors.push(format!("Invalid break strength: {}", strength));
                }
            }
        }
    }
}

/// Predefined SSML templates
pub struct SSMLTemplates;

impl SSMLTemplates {
    /// Create SSML using a predefined template
    pub fn create_from_template(template_name: &str, text: &str, voice: &str) -> Result<String, String> {
        match template_name {
            "slow_speech" => Ok(SSMLBuilder::new(voice).add_prosody(text, Some("slow"), None, None).build()),
            "fast_speech" => Ok(SSMLBuilder::new(voice).add_prosody(text, Some("fast"), None, None).build()),
            "whisper" => Ok(SSMLBuilder::new(voice).add_prosody(text, Some("slow"), None, Some("x-soft")).build()),
            "excited" => Ok(SSMLBuilder::new(voice).add_prosody(text, Some("fast"), Some("high"), Some("loud")).build()),
            "calm" => Ok(SSMLBuilder::new(voice).add_prosody(text, Some("slow"), Some("low"), Some("soft")).build()),
            "emphasis_strong" => Ok(SSMLBuilder::new(voice).add_emphasis(text, "strong").build()),
            "with_pauses" => {
                if text.contains('.') {
                    let parts: Vec<&str> = text.split('.').collect();
                    if parts.len() >= 2 {
                        Ok(SSMLBuilder::new(voice)
                            .add_text(parts[0])
                            .add_break("1s")
                            .add_text(&parts[1..].join("."))
                            .build())
                    } else {
                        Ok(SSMLBuilder::new(voice).add_text(text).build())
                    }
                } else {
                    Ok(SSMLBuilder::new(voice).add_text(text).build())
                }
            },
            _ => {
                let available = "slow_speech, fast_speech, whisper, excited, calm, emphasis_strong, with_pauses";
                Err(format!("Unknown template '{}'. Available: {}", template_name, available))
            }
        }
    }

    /// Get list of available template names
    pub fn get_available_templates() -> Vec<&'static str> {
        vec![
            "slow_speech",
            "fast_speech", 
            "whisper",
            "excited",
            "calm",
            "emphasis_strong",
            "with_pauses"
        ]
    }
}

/// Validate SSML markup
pub fn validate_ssml(ssml: &str, raise_on_error: bool) -> Result<Vec<String>, String> {
    let errors = SSMLValidator::validate(ssml);
    
    if !errors.is_empty() && raise_on_error {
        return Err(format!("SSML validation failed: {}", errors.join("; ")));
    }
    
    Ok(errors)
}

/// Create SSML with prosody controls
pub fn create_ssml(text: &str, voice: &str, rate: Option<&str>, pitch: Option<&str>, volume: Option<&str>) -> String {
    SSMLBuilder::new(voice).add_prosody(text, rate, pitch, volume).build()
}

/// Create SSML with emphasis
pub fn create_emphasis_ssml(text: &str, voice: &str, level: &str) -> String {
    SSMLBuilder::new(voice).add_emphasis(text, level).build()
}

/// Create SSML with breaks between text parts
pub fn create_break_ssml(text_parts: &[&str], voice: &str, break_time: &str) -> String {
    let mut builder = SSMLBuilder::new(voice);
    
    for (i, part) in text_parts.iter().enumerate() {
        builder = builder.add_text(part);
        if i < text_parts.len() - 1 {
            builder = builder.add_break(break_time);
        }
    }
    
    builder.build()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ssml_builder_basic() {
        let ssml = SSMLBuilder::new("en-US-AriaNeural")
            .add_text("Hello, world!")
            .build();
        
        assert!(ssml.contains("<speak"));
        assert!(ssml.contains("en-US-AriaNeural"));
        assert!(ssml.contains("Hello, world!"));
    }

    #[test]
    fn test_ssml_builder_prosody() {
        let ssml = SSMLBuilder::new("en-US-AriaNeural")
            .add_prosody("Hello", Some("slow"), Some("high"), Some("loud"))
            .build();
        
        assert!(ssml.contains("rate=\"slow\""));
        assert!(ssml.contains("pitch=\"high\""));
        assert!(ssml.contains("volume=\"loud\""));
    }

    #[test]
    fn test_ssml_builder_emphasis() {
        let ssml = SSMLBuilder::new("en-US-AriaNeural")
            .add_emphasis("Important!", "strong")
            .build();
        
        assert!(ssml.contains("<emphasis level=\"strong\">Important!</emphasis>"));
    }

    #[test]
    fn test_ssml_builder_break() {
        let ssml = SSMLBuilder::new("en-US-AriaNeural")
            .add_text("First part")
            .add_break("2s")
            .add_text("Second part")
            .build();
        
        assert!(ssml.contains("<break time=\"2s\"/>"));
    }

    #[test]
    fn test_ssml_validation_valid() {
        let ssml = SSMLBuilder::new("en-US-AriaNeural")
            .add_text("Hello")
            .build();
        
        let errors = SSMLValidator::validate(&ssml);
        assert!(errors.is_empty());
    }

    #[test]
    fn test_ssml_validation_invalid() {
        let invalid_ssml = "<invalid>test</invalid>";
        let errors = SSMLValidator::validate(invalid_ssml);
        assert!(!errors.is_empty());
    }

    #[test]
    fn test_templates() {
        let result = SSMLTemplates::create_from_template("slow_speech", "Hello", "en-US-AriaNeural");
        assert!(result.is_ok());
        assert!(result.unwrap().contains("rate=\"slow\""));
    }

    #[test]
    fn test_template_unknown() {
        let result = SSMLTemplates::create_from_template("unknown", "Hello", "en-US-AriaNeural");
        assert!(result.is_err());
    }
}